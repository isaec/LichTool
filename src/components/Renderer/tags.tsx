import { fmtDataUrl } from "@src/formatter";
import { Link } from "solid-app-router";
import { Component } from "solid-js";
import { JSX } from "solid-js/jsx-runtime";
import { Dynamic } from "solid-js/web";

import styles from "./Renderer.module.scss";

/**
 * maps alias shorthand "b" to full tag "bold"
 */
export const tagAlias = new Map([
  ["b", "bold"],
  ["i", "italic"],
  ["u", "underline"],
  ["s", "strike"],
]);

const pipe =
  (Element: Component<{ p: string[]; p0: string; p1?: string; p2?: string }>) =>
  (props: { children: JSX.Element }): JSX.Element => {
    const pipeSplit =
      typeof props.children === "string" ? props.children.split("|") : [""];
    return (
      <Element
        p={pipeSplit}
        p0={pipeSplit[0]}
        p1={pipeSplit[1]}
        p2={pipeSplit[2]}
      />
    );
  };

const refTag = (type: string, defaultSource: string) =>
  pipe((props) => (
    <Link
      href={`/view/${fmtDataUrl(
        type,
        props.p0,
        props.p1 === undefined || props.p1 === "" ? defaultSource : props.p1
      )}`}
    >
      {props.p2 ?? props.p0}
    </Link>
  ));

const tag =
  (
    tagString: keyof JSX.IntrinsicElements,
    obj?: {
      class?: string;
      style?: string;
    }
  ): Component<{ children: JSX.Element }> =>
  (props) =>
    (
      <Dynamic
        component={tagString}
        children={props.children}
        class={obj?.class}
        style={obj?.style}
      />
    );

const templateTag =
  (
    tagString: keyof JSX.IntrinsicElements,
    templater: (arg0: JSX.Element) => JSX.Element,
    obj?: {
      class?: string;
      style?: string;
    }
  ): Component<{ children: JSX.Element }> =>
  (props) =>
    (
      <Dynamic
        component={tagString}
        children={templater(props.children)}
        class={obj?.class}
        style={obj?.style}
      />
    );

export const tagMap = new Map<string, Component<{ children: JSX.Element }>>(
  Object.entries({
    // intrinsic tags
    bold: tag("b"),
    italic: tag("i"),
    strike: tag("s"),
    code: tag("code"),
    // styled intrinsic tags
    underline: tag("span", { class: styles.underline }),
    note: tag("i", { class: styles.note }),
    // complex tags
    color: pipe((props) => (
      <span style={`color: #${props.p1}`}>{props.p0}</span>
    )),
    link: pipe((props) => (
      <a href={props.p1 ?? props.p0} target="_blank" rel="noopener noreferrer">
        {props.p0}
      </a>
    )),
    // link tags
    spell: refTag("spell", "phb"),
    condition: refTag("condition", "phb"),
    action: refTag("action", "phb"),
    // bad / unsupported tags
    filter: pipe((props) => props.p0),
    dice: tag("code"),
    hit: templateTag("code", (c) => <>d20{c}</>),
    damage: tag("code"),
    d20: templateTag("code", (c) => <>d20{c}</>),
    scaledice: pipe((props) => <code>{props.p2}</code>),
    scaledamage: pipe((props) => <code>{props.p2}</code>),
  })
);
