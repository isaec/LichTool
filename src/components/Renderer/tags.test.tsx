import { Dynamic } from "solid-js/web";
import { render } from "solid-testing-library";
import { vi, expect, describe, it, test } from "vitest";
import styles from "./Renderer.module.scss";
import { tagMap } from "./tags";

vi.mock("./Renderer.module.scss", () => ({
  default: new Proxy(new Object(), {
    get(_, style) {
      return style;
    },
  }),
}));

describe("tags", () => {
  const needPipe = new Set(["color", "link", "filter"]);
  it.each([
    ["color", "test|ffffff"],
    ["color", "test|f00"],
    ["color", "test|illegal-color"],
    ["color", "test|fff|bonus-pipe"],
    ...[...tagMap.keys()]
      .filter((tag) => !needPipe.has(tag))
      .map((key) => [key, `${key} string!`]),
  ])(`{@%s %s} matches snapshot`, (tag, children) => {
    const { unmount, container } = render(() => (
      <Dynamic component={tagMap.get(tag)} children={children} />
    ));

    expect(container).toMatchSnapshot();
    unmount();
  });
});
