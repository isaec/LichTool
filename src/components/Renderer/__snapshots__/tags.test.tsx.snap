// Vitest Snapshot v1

exports[`tags > {@bold bold string!} matches snapshot 1`] = `
<div>
  <b>
    bold string!
  </b>
</div>
`;

exports[`tags > {@code code string!} matches snapshot 1`] = `
<div>
  <code>
    code string!
  </code>
</div>
`;

exports[`tags > {@color test|f00} matches snapshot 1`] = `
<div>
  <span
    style="color: rgb(255, 0, 0);"
  >
    test
  </span>
</div>
`;

exports[`tags > {@color test|fff|bonus-pipe} matches snapshot 1`] = `
<div>
  <span
    style="color: rgb(255, 255, 255);"
  >
    test
  </span>
</div>
`;

exports[`tags > {@color test|ffffff} matches snapshot 1`] = `
<div>
  <span
    style="color: rgb(255, 255, 255);"
  >
    test
  </span>
</div>
`;

exports[`tags > {@color test|illegal-color} matches snapshot 1`] = `
<div>
  <span
    style=""
  >
    test
  </span>
</div>
`;

exports[`tags > {@italic italic string!} matches snapshot 1`] = `
<div>
  <i>
    italic string!
  </i>
</div>
`;

exports[`tags > {@note note string!} matches snapshot 1`] = `
<div>
  <i
    class="note"
  >
    note string!
  </i>
</div>
`;

exports[`tags > {@strike strike string!} matches snapshot 1`] = `
<div>
  <s>
    strike string!
  </s>
</div>
`;

exports[`tags > {@underline underline string!} matches snapshot 1`] = `
<div>
  <span
    class="underline"
  >
    underline string!
  </span>
</div>
`;
